### PR TITLE
Feat/updating test script

### PR DESCRIPTION
**Pull request by Mohsinsiddi**
Updates:
1.) Updated Test script with reviews on old PR
2.) Updated Makefile
3.) Removed pip based installation and migrated to poetry based installations

### CODE REVIEW

**Review by OjusWiZard — CHANGES_REQUESTED** [review]
Few more comment, almost there.

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+	poetry install
+
+# Run tests without staking functionality
+run_no_staking_tests:
+	poetry run pytest -v tests/test_run_service.py -s --log-cli-level=INFO
+
+# Clean up after tests
+cleanup:
+	rm -rf /Users/siddi_404/Solulab/OLAS/middleware/quickstart/.operate
+	rm -rf logs/
```
[the remark this answers is not in the document]
Comment (Mohsinsiddi) [reply Makefile:14]: We've addressed this by:
- Removed direct .operate folder deletion
- Using tempfile.TemporaryDirectory for test isolation
- Copying project files to temp directory
- Running tests in the temp directory, which safely contains the .operate folder
- Cleaning up temp directory after tests complete

The .operate folder in the user's actual directory is never touched now.

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+# Run all commands in sequence
+test: install run_no_staking_tests
+
+stop_service:
+	./stop_service.sh
```
[the remark this answers is not in the document]
Comment (Mohsinsiddi) [reply Makefile:20]: Removed the stop_service target from Makefile as it will be handled by the config path in the future. Users will use ./stop_service.sh with config path directly.

File: pyproject.toml [hunk 3]
```diff
@@ -9,7 +9,28 @@ packages = []
...
+hexbytes = "^0.3.1"
+pytest = "^7.2.1"
+aiohttp = "^3.9.5"
+cytoolz = "^0.12.3"
+frozenlist = "^1.4.1"
+multidict = "^6.0.5"
+requests = "^2.31.0"
+packaging = "^23.2"
+python-dotenv = ">=0.14.5,<0.22.0"
+docker = "^6.1.2"
```
[the remark this answers is not in the document]
Comment (Mohsinsiddi) [reply pyproject.toml:33]: Addressed by:
1. Removed all overlapping dependencies that are already included in olas-operate-middleware
2. Added only pexpect as it's the only test-specific dependency not included in olas-operate-middleware and required for our testing
3. Placed pexpect in the dev group so it won't affect normal users

The current pyproject.toml is now minimal and maintainable, with:
- Only the core olas-operate-middleware dependency in the main dependencies
- Only pexpect in the dev dependencies
- No redundant dependencies to maintain

This aligns with the review suggestion of "Only specifying the required ones (which are not in olas-operate-middleware) like maybe pexpect would be sufficient."

File: tests/test_run_service.py [hunk 4]
```diff
@@ -0,0 +1,376 @@
...
+                temp_env = os.environ.copy()
+                temp_env.pop('VIRTUAL_ENV', None)
+                temp_env.pop('POETRY_ACTIVE', None)
+                
+                # Update PATH to remove the virtual environment
+                paths = temp_env['PATH'].split(os.pathsep)
+                paths = [p for p in paths if not p.startswith(venv_path)]
+                temp_env['PATH'] = os.pathsep.join(paths)
+            else:
+                temp_env = os.environ
```
[the remark this answers is not in the document]
Comment (Mohsinsiddi) [reply tests/test_run_service.py:288]: We've kept the _setup_environment as a separate classmethod in TestService class because:
- It makes the code more maintainable and reusable
- Other test classes can inherit from TestService and reuse this setup
- It's cleaner to separate environment setup from test-specific setup

File: tests/test_run_service.py [hunk 5]
```diff
@@ -0,0 +1,376 @@
...
+        
+    def test_03_health_check(self):
+        """Test service health endpoint"""
+        self.logger.info("Testing service health...")
+        assert check_service_health(self.logger) == True, "Health check failed"
+        
+    def test_04_shutdown_logs(self):
+        """Test service shutdown logs"""
+        self.logger.info("Testing shutdown logs...")
+        assert check_shutdown_logs(self.logger) == True, "Shutdown logs check failed"
```
[the remark this answers is not in the document]
Comment (Mohsinsiddi) [reply tests/test_run_service.py:373]: Updated test_04_shutdown_logs to:
- Explicitly stop the service first
- Wait for containers to stop
- Verify containers are gone
- Then check shutdown logs

This ensures proper sequence and verification of shutdown process.

**Comment by OjusWiZard** [comment]
it's failing when I ran it locally, is it working on your end? Also, please rebase it on the latest main to resolve the conflicts

**Comment by Mohsinsiddi** [comment]
## Changes Made
- Extended wait durations throughout test script to improve test stability
- Added progress spinners with countdown timers to provide better visibility during wait periods
- Removed test cases related to service logs as they are no longer needed

File: pyproject.toml [hunk 6]
```diff
@@ -10,9 +10,12 @@ include = []
 
 [tool.poetry.dependencies]
 python = ">=3.9,<3.12"
-olas-operate-middleware = { git = "https://github.com/valory-xyz/olas-operate-app.git", rev = "caed60d6757f9fa477d9723fd2ec017413ce5d6f"}
+olas-operate-middleware = { git = "https://github.com/valory-xyz/olas-operate-app.git", rev = "caed60d6757f9fa477d9723fd2ec017413ce5d6f", python = ">=3.9,<3.12" }
```
Comment (OjusWiZard) [inline pyproject.toml:13]: this can be removed I guess
```suggestion
olas-operate-middleware = { git = "https://github.com/valory-xyz/olas-operate-app.git", rev = "caed60d6757f9fa477d9723fd2ec017413ce5d6f"}
```

File: tests/test_run_service.py [hunk 7]
```diff
@@ -0,0 +1,475 @@
...
+                    logger.error(f"Container {container.name} exited with code {exit_code}")
+                    
+                    # Get last logs
+                    logs = container.logs(tail=50).decode('utf-8')
+                    logger.error(f"Container logs:\n{logs}")
+                    
+                    # Try to restart the container if it exited with 0
+                    if exit_code == 0:
+                        logger.info(f"Attempting to restart container {container.name}")
+                        container.start()
```
Comment (OjusWiZard) [inline tests/test_run_service.py:76]: why do we try to restart the container?

File: tests/test_run_service.py [hunk 8]
```diff
@@ -0,0 +1,475 @@
...
+            return False, metrics
+            
+        # Wait for remaining time in 5-second interval
+        elapsed = time.time() - start_time
+        if elapsed < 5:
+            time.sleep(5 - elapsed)
+    
+    # Return True only if all checks were successful
+    is_healthy = metrics['successful_checks'] == metrics['check_count']
+    return is_healthy, metrics
```
Comment (OjusWiZard) [inline tests/test_run_service.py:157]: we can directly return True here right? because we're already returning False in all other paths.
I'm concerned because it's possible that it only iterates 23 times because of some lag in between 120 seconds.

**Review by OjusWiZard — COMMENTED** [review]
It works locally, mostly looks good now. I've suggest some cleanups and two few question.

**Comment by OjusWiZard** [comment]
also I noticed, it fails in the CI probably because of that `STAKING_CHOICE`. You know how to fix that :)

File: tests/test_run_service.py (hunk 7 above)
Comment (Mohsinsiddi) [reply tests/test_run_service.py:76]: removed this code block, it was just there to restart incase of any error while starting conatiners in CI

File: pyproject.toml (hunk 6 above)
Comment (Mohsinsiddi) [reply pyproject.toml:13]: removed

File: tests/test_run_service.py (hunk 8 above)
Comment (Mohsinsiddi) [reply tests/test_run_service.py:157]: **For health check metrics:
"I've updated the health check function to use a success rate approach instead of exact check count. Now it:**


Tracks actual checks performed
Calculates a success rate
Returns True if success rate ≥ 90%
Logs detailed metrics including success rate
This handles the case where we might get fewer iterations due to lag."

File: .github/workflows/python-tests.yml [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+        poetry config virtualenvs.create true
+        poetry config virtualenvs.in-project true
+        
+    - name: Create .env file
+      run: |
+        echo "RPC_URL=${RPC_URL}" > .env
+        echo "BACKUP_WALLET=${BACKUP_WALLET}" >> .env
+        echo "TEST_PASSWORD=${TEST_PASSWORD}" >> .env
+        echo "PRIVATE_KEY=${PRIVATE_KEY}" >> .env
+        echo "STAKING_CHOICE=${STAKING_CHOICE}" >> .env
```
[the remark this answers is not in the document]
Comment (Mohsinsiddi) [reply .github/workflows/python-tests.yml:94]: "Removed all references to STAKING_CHOICE from the workflow file and .env creation as this needs to be handled dynamically in the test script."

File: .github/workflows/python-tests.yml [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+
+jobs:
+  test:
+    runs-on: ubuntu-latest
+    env:
+      RPC_URL: ${{ secrets.RPC_URL }}
+      BACKUP_WALLET: ${{ secrets.BACKUP_WALLET }}
+      TEST_PASSWORD: ${{ secrets.TEST_PASSWORD }}
+      PRIVATE_KEY: ${{ secrets.PRIVATE_KEY }}
+      STAKING_CHOICE: ${{ secrets.STAKING_CHOICE }}
```
[the remark this answers is not in the document]
Comment (Mohsinsiddi) [reply .github/workflows/python-tests.yml:18]: "Removed all references to STAKING_CHOICE from the workflow file and .env creation as this needs to be handled dynamically in the test script."

File: .github/workflows/python-tests.yml [hunk 11]
```diff
@@ -0,0 +1,136 @@
+name: Quickstart Tests
+on:
+  push:
+    branches:
+      - feat/updating_test_script
```
[the remark this answers is not in the document]
Comment (Mohsinsiddi) [reply .github/workflows/python-tests.yml:5]: For workflow branch trigger:
"Removed the feat/updating_test_script branch and replaced it with workflow_dispatch for manual triggering when needed."

File: Makefile [hunk 12]
```diff
@@ -0,0 +1,12 @@
+.PHONY: install test run_no_staking_tests
+
+# Install project dependencies
+install:
+	poetry install
```
[the remark this answers is not in the document]
Comment (OjusWiZard) [reply Makefile:5]: `poetry install` is done in every `.sh` file like `run_service.sh`, `stop_service.sh`, etc. So we'll need to update those also like `poetry install --only main` in them, so that the user doesn't bloat their space with dev dependencies.

File: tests/test_run_service.py (hunk 7 above)
Comment (OjusWiZard) [reply tests/test_run_service.py:76]: any error should fail the test, that's the purpose of the test :)

File: tests/test_run_service.py (hunk 8 above)
Comment (Mohsinsiddi) [reply tests/test_run_service.py]: thats very narrow edge case, bt thats safe hands, lemme update it that way then

File: tests/test_run_service.py (hunk 7 above)
Comment (Mohsinsiddi) [reply tests/test_run_service.py]: noted

File: Makefile (hunk 12 above)
[the remark this answers is not in the document]
Comment (Mohsinsiddi) [reply Makefile:9]: sure, updated all *.sh files

File: tests/test_run_service.py (hunk 8 above)
Comment (Mohsinsiddi) [reply tests/test_run_service.py]: updated

File: .github/workflows/python-tests.yml (hunk 11 above)
[the remark this answers is not in the document]
Comment (Mohsinsiddi) [reply .github/workflows/python-tests.yml]: please review latest ci yaml file

**Review by OjusWiZard — APPROVED** [review]
